### PR TITLE
flax model parallel training

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

Adds model parallel training example for GPTNeo using jax's [`pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html) transformation.
(This example probably just works on a single TPU v3-8).

This should enable training bigger models like 1.3B GPTNeo on a single TPU V3-8.

The `partition.py` file defines the `PyTree` of the `PartitionSpec` file which describes how the model parameters will be sharded. The actual sharding is automatically handled by `pjit`. 

The key idea is to `pjit` the entire training step function. To do that we
- Define the mesh structure.
- Define `PartitionSpec` for every input argument and return value of the pjitted function. The axis names that are used here should match the axis names used in `PartitionSpec`. This means we need the spec for our parameter and optimizer state PyTrees
- The structure of the `PyTree` of `PartitionSpec` needs to match the structure of the `PyTree` of the actual values.
- Call the pijitted fun in a mesh context.

Below is a not-so minimal code-snippet that describes the approach

```python

# init our model
model = FlaxGPTNeoForCausalLM.from_pretrained("gpt-neo-125M")

# get the partition spec for model params
param_spec = set_partitions(unfreeze(model.params))

# get optimizer
optim = optax.adamw(learning_rate=decay_fn)

# mesh defination
mesh_devices = np.array(jax.devices()).reshape(1, jax.local_device_count())

def get_initial_state(params):
    state = optim.init(params)
    return tuple(state), params

# init optim in abstract way, this just returns the PyTree of opt_state with shapes
# so we can get the PartitionSpec for opt_state using the tree
shapes = jax.tree_map(lambda x: x.shape, model.params)
state = jax.eval_shape(get_initial_state, shapes)

# Get the opt spec
def get_opt_spec(x):
    if isinstance(x, dict):
        return param_spec
    return None

opt_state_spec, param_spec = jax.tree_map(
    get_opt_spec, state, is_leaf=lambda x: isinstance(x, (dict, optax.EmptyState))
)

# Now actually initialize the opt state
# this also takes care of sharding the opt and param state according to the spec.
p_get_initial_state = pjit(
    get_initial_state,
    in_axis_resources=None,
    out_axis_resources=(opt_state_spec, param_spec),
)
with mesh(mesh_devices, ("dp", "mp")):
    opt_state, params = p_get_initial_state(freeze(model.params))

# define out train step
def train_step(params, opt_state, dropout_rng, batch):
	....
	return new_params, tuple(new_opt_state), new_dropout_rng, metrics

# pjit the train step
# in_axis_resources and out_axis_resources expect the PartitionSpec 
# for every input argument and return values
p_train_step = pjit(
    train_step,
    in_axis_resources=(param_spec, opt_state_spec, None, None),
    out_axis_resources=(param_spec, opt_state_spec, None, None),
)

# do the training
with mesh(mesh_devices, ("dp", "mp")):
    params, state, loss, rng = p_train_step(params, opt_state, ...)
```

As we can see above, all the sharding logic is outside of the model definition, so ideally we don't need to modify the modeling code. This also means it should be possible to apply this to any other model by defining the right `PyTree` of `PartitionSpec`. 

A few things to consider for future work.

- A convenient way to get the PyTree of model parameters, so we can define the partition spec.
- Currently,  model weights are initialized when the model class is instantiated. This can cause problems for models that cannot fit on one device. There should be an option to abstractly initialize the model without having to initialize the weights. This will also allow a convenient way to get the PyTree.
- The `from_pretrained` method also directly puts the weights on the device, we need to consider either sharded loading or initially loading the weights on the CPU then sharding them on the devices, to avoid OOM with huge models.

### CODE REVIEW

File: examples/research_projects/jax-projects/model parallel/run_clm_mp.py [hunk 1]
```diff
@@ -0,0 +1,648 @@
...
+    logger.info(f"  Instantaneous batch size per device = {training_args.per_device_train_batch_size}")
+    logger.info(f"  Total train batch size (w. parallel & distributed) = {train_batch_size}")
+    logger.info(f"  Total optimization steps = {total_train_steps}")
+
+    train_time = 0
+    train_metrics = []
+    epochs = tqdm(range(num_epochs), desc=f"Epoch ... (1/{num_epochs})", position=0)
+    global_step = 0
+    # we are not doing 2D parallelism (yet!), this just does model parallelism
+    with mesh(mesh_devices, ("dp", "mp")):
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/model parallel/run_clm_mp.py:572]: Why do we define "dp" here? It's not used in any partition spec no?

File: examples/research_projects/jax-projects/model parallel/partitions.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            if _match(rule, key):
+                return replacement
+        return val
+
+    return replace
+
+
+# PartitionSpec for GPTNeo
+# replicate the hidden dim and shard feed-forward and head dim
+def _get_partition_rules():
```
Comment (patrickvonplaten) [inline examples/research_projects/jax-projects/model parallel/partitions.py:57]: For the future:

Do you think we'll always need model-specific partitions? This might make it a bit difficult to implement for all models, no? I guess we can't just assign the correct `P("mp", None)` depending on the last word "kernel" or "embedding"  since they switch...

Maybe it's somehow possible to trace a model (similar to how init does in to give us the params) to automatically build this map?

**Review by patrickvonplaten — APPROVED** [review]
Awesome! Think for the future it would be great to find a way to automatically generated a model partition map by tracing a dummy model somehow (maybe with a function similar to `init`). Not sure...it's probably not that easy...

 Also cc @stas00 here might be interesting :-)

File: examples/research_projects/jax-projects/model parallel/partitions.py (hunk 2 above)
Comment (patil-suraj) [reply examples/research_projects/jax-projects/model parallel/partitions.py:57]: That's the idea. With lazy initialization, we get the `PyTree` of params which describes all param names and shapes. 
With that, it should be possible to build a default spec for every model. 

We can already do this with `jax.eval_shape`. cf
https://github.com/huggingface/transformers/blob/399258357bad74186f248a373af8012f2b0203b7/examples/research_projects/jax-projects/model%20parallel/run_clm_mp.py#L471

Comment (stas00) [reply examples/research_projects/jax-projects/model parallel/partitions.py:57]: This is typically done using using various load balancing methods, e.g. deepspeed pipe has:
https://www.deepspeed.ai/tutorials/pipeline/#load-balancing-pipeline-modules
pytorch has these too but I can't find any mentions of these in their docs.

Is that what you're referring to, @patrickvonplaten

**Comment by stas00** [comment]
This is typically done using using various load balancing methods, e.g. deepspeed pipe has:
https://www.deepspeed.ai/tutorials/pipeline/#load-balancing-pipeline-modules

pytorch has these too but I can't find any mentions of these in their docs. 
Have to go to the source:
https://github.com/pytorch/pytorch/blob/58adaaba60441c1ed59f35389598aabf91a772dd/torch/distributed/pipeline/sync/_balance/__init__.py

Is that what you're referring to, @patrickvonplaten
